cmd/collector/app/metrics.go

```diff
@@ -169,19 +166,30 @@ func (m metricsBySvc) countTracesByServiceName(serviceName string) {
 // total number of stored counters, so if it exceeds say the 90% threshold
 // an alert should be raised to investigate what's causing so many unique
 // service names.
-func (m *countsBySvc) countByServiceName(serviceName string) {
+func (m *countsBySvc) countByServiceName(serviceName string, isDebug bool) {
 	serviceName = NormalizeServiceName(serviceName)
+	getOrCreateCounter := func(counts map[string]metrics.Counter, createCounter func() metrics.Counter) metrics.Counter {
+		if c, ok := counts[serviceName]; ok {
+			return c
+		}
+		if len(counts) < m.maxServiceNames {
+			c := createCounter()
+			counts[serviceName] = c
+			return c
+		}
+		return counts[otherServices]
+	}
 	var counter metrics.Counter
 	m.lock.Lock()
-	if c, ok := m.counts[serviceName]; ok {
-		counter = c
-	} else if len(m.counts) < maxServiceNames {
-		c := m.factory.Counter("", map[string]string{"service": serviceName})
-		m.counts[serviceName] = c
-		counter = c
+	if isDebug {
+		counter = getOrCreateCounter(m.debugCounts, func() metrics.Counter {
+			return m.factory.Counter("", map[string]string{"service": serviceName, "debug": "true"})
+		})
+	} else {
+		counter = getOrCreateCounter(m.counts, func() metrics.Counter {
+			return m.factory.Counter("", map[string]string{"service": serviceName, "debug": "false"})
+		})
```

Comment:
this is highly in efficient. The overall function is "count", one would expect it to be quick and zero-allocations, but you are creating all these lambdas that will require memory allocations.

Reply:
went from `1067 ns/op	     231 B/op	      11 allocs/op` to `947 ns/op	     165 B/op	       9 allocs/op`

Reply:
maybe if you benchmark the unhappy path, but try with the same service name in a hot loop